### PR TITLE
Split up matchmaker_config.yaml

### PR DESCRIPTION
**Pull request by ihrankouski**
This proposes a solution to #130.

The configuration of core Open-match components ('services') is split into 3 files:
- /config/component_config.yaml - contains a per-component configuration
- /config/openmatch_config.yaml - configuration that must be shared by all components
- /config/openmatch_constants.yaml - configuration that must be shared but is more dangerous to modify (so it's built into docker images)

File "openmatch_config.yaml" is also mounted to MMF and Evaluator Job's containers.

Original "matchmaker_config.yaml" is removed along with all its symlinks.

Also moved config package under internal/.

Also I placed the contents of "component_config.yaml" and "openmatch_config.yaml" inside the values.yaml file of Helm chart. So in theory one could even override some parts of Open-match configuration at chart installation time (didn't test how it works yet).

Note that configuration watching and reloading is not implemented at the moment. (Because of merging of several files in Viper and usage of "subPath")

### CODE REVIEW

File: config/openmatch_constants.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# A copy of this file is built into Open-match components' Docker images.
```
[the remark this answers is not in the document]
Comment (ihrankouski) [reply config/openmatch_constants.yaml]: On the other side if we always keep all configuration in ConfigMap it will be easier to get config reloading back...

File: test/cmd/clientloadgen/main.go [hunk 2]
```diff
@@ -23,15 +23,15 @@ import (
...
+	"github.com/GoogleCloudPlatform/open-match/internal/config"
 	"github.com/GoogleCloudPlatform/open-match/test/cmd/clientloadgen/player"
 	"github.com/GoogleCloudPlatform/open-match/test/cmd/clientloadgen/redis/playerq"
 
 	"github.com/gomodule/redigo/redis"
 )
 
 func main() {
-	cfg, err := config.Read()
+	cfg, err := config.Read("config.yaml")
```
[the remark this answers is not in the document]
Comment (ihrankouski) [reply test/cmd/clientloadgen/main.go]: Makes sense. 'clientloadgen' is just a tool to flood the state storage with dummy data, and it needed it's own configuration (that is not related to core components' params). I agree, having a convention of ['config.yaml' + ReadConfig()] would be sufficient to make pkg /internal/config more reusable, will add that.

File: examples/backendclient/main.go [hunk 3]
```diff
@@ -138,6 +137,7 @@ func main() {
 				break
 			}
 
+			match := listResponse.Match
```
Comment (jeremyje) [inline examples/backendclient/main.go]: Can you sync? This was fixed.

File: examples/functions/python3/mmlogic-simple/harness.py [hunk 4]
```diff
@@ -28,13 +29,13 @@
...
 # Establish grpc channel and make the API client stub
-api_conn_info = "%s:%d" % (os.environ["OM_MMLOGICAPI_SERVICE_HOST"],os.environ["OM_MMLOGICAPI_SERVICE_PORT"])
+api_conn_info = "%s:%s" % (os.environ["OM_MMLOGICAPI_SERVICE_HOST"],os.environ["OM_MMLOGICAPI_SERVICE_PORT"])
 with  grpc.insecure_channel(api_conn_info) as channel:
     mmlogic_api = mmlogic_pb2_grpc.MmLogicStub(channel)
 
     # Step 3 - Read the profile written to the Backend API.
     # Get profile from redis
-    profile_pb = mmlogic_api.GetProfile(mmlogic.MatchObject(id=os.environ["MMF_PROFILE_ID"]))
+    profile_pb = mmlogic_api.GetProfile(mmlogic.GetProfileRequest(match = pb.MatchObject(id=os.environ["MMF_PROFILE_ID"]))).match
```
Comment (jeremyje) [inline examples/functions/python3/mmlogic-simple/harness.py]: Split this into a separate change. Much appreciated fix.

Comment (ihrankouski) [reply examples/functions/python3/mmlogic-simple/harness.py]: done #193

File: examples/backendclient/main.go (hunk 3 above)
Comment (ihrankouski) [reply examples/backendclient/main.go]: done

File: examples/backendclient/main.go [hunk 5]
```diff
@@ -138,6 +138,7 @@ func main() {
 			}
 			match := listResponse.Match
 
+			match := listResponse.Match
```
Comment (jeremyje) [inline examples/backendclient/main.go]: Bad merge?

File: test/cmd/clientloadgen/Dockerfile [hunk 6]
```diff
@@ -17,8 +17,9 @@ FROM open-match-base-build as builder
 WORKDIR /go/src/github.com/GoogleCloudPlatform/open-match/test/cmd/clientloadgen/
 RUN CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o clientloadgen . 
 
-FROM gcr.io/distroless/static  
+FROM alpine
```
Comment (jeremyje) [inline test/cmd/clientloadgen/Dockerfile]: Why?

File: test/cmd/clientloadgen/Dockerfile [hunk 7]
```diff
@@ -17,8 +17,9 @@ FROM open-match-base-build as builder
...
 COPY --from=builder /go/src/github.com/GoogleCloudPlatform/open-match/test/cmd/clientloadgen/clientloadgen .
+COPY --from=builder /go/src/github.com/GoogleCloudPlatform/open-match/test/cmd/clientloadgen/config.yaml ./
```
Comment (jeremyje) [inline test/cmd/clientloadgen/Dockerfile]: Please don't add configs to the images.

File: examples/backendclient/main.go (hunk 5 above)
Comment (ihrankouski) [reply examples/backendclient/main.go]: ooops...

File: internal/config/merged_config.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+	for _, l := range layers {
+		m := l.AllSettings()
+		cfg.MergeConfigMap(m)
+	}
+	return cfg
+}
+
+// Wrapper struct that implements View interface
+// and delegates to other viper.Viper instance
+type wrapperView struct {
```
Comment (jeremyje) [inline internal/config/merged_config.go]: Why is this necessary if you have the merge() function?

File: internal/config/view.go [hunk 9]
```diff
@@ -34,6 +34,7 @@ type View interface {
 	GetBool(string) bool
 	GetDuration(string) time.Duration
 	GetStringMap(string) map[string]interface{}
+	AllSettings() map[string]interface{}
```
Comment (jeremyje) [inline internal/config/view.go]: AllSettings is only used against concrete `*viper.Viper` so this isn't necessary. If you need it because of tests I'd prefer a type assertion case to viper.Viper there since you already know you're testing against that class to begin with.

**Comment by jeremyje** [comment]
Those should be my last comments on this change.

File: internal/config/merged_config.go (hunk 8 above)
Comment (ihrankouski) [reply internal/config/merged_config.go]: I didn't find a function to replace all contents of a Viper instance. That means that if we use the same instance upon layers' reloads, then that instance might become polluted with some old data from previous reads.

File: internal/config/view.go (hunk 9 above)
Comment (ihrankouski) [reply internal/config/view.go]: Makes sense.

**Comment by ihrankouski** [comment]
Well `go test` reports a data race: https://console.cloud.google.com/cloud-build/builds/c2a4b933-405e-4440-b52a-e6d21c872c6c?project=open-match-build#step_15
I'm not sure about how to deal with that because the issue seems to originate inside Viper, similar issues:
https://github.com/spf13/viper/issues/174
https://github.com/spf13/viper/issues/268

File: internal/config/view.go (hunk 9 above)
Comment (ihrankouski) [reply internal/config/view.go]: Done, thanks!

**Comment by jeremyje** [comment]
Were you able to get rid of the race condition?
I think for the purpose of the change we should break out the file move first and then add the split if the race condition still is there.

**Comment by jeremyje** [comment]
I think the easiest thing to do is bring back the wrapper and add a lock to use each method. We don't read from the config often enough for it to hurt perf.

**Comment by jeremyje** [comment]
Merge conflicts. CI via Make just got submitted should provide better regression coverage.

**Comment by ihrankouski** [comment]
Okay I'm still fighting with data races. What's weird is that "go1.12.1 darwin/amd64" does not report races on this branch for me any more (after my attempt to fix races). However I sometimes can reproduce it on clean instances in GCE.
I'll open a new PR with only files move.

**Comment by jeremyje** [comment]
https://github.com/GoogleCloudPlatform/open-match/issues/213

**Comment by jeremyje** [comment]
There's a lot of conflicts here. Is this change still going in?

**Comment by ihrankouski** [comment]
I'm closing this, let's try to get #204 working at first.
